core-rust/state-manager/src/staging/result.rs

```diff
@@ -264,6 +245,35 @@ impl ProcessedCommitResult {
             .map(|next_epoch_result| NextEpoch::from(next_epoch_result.clone()))
     }
 
+    pub fn compute_substate_changes<S: SubstateDatabase, D: DatabaseKeyMapper>(
+        store: &S,
+        system_updates: &SystemUpdates,
+    ) -> Vec<SubstateChange> {
+        let mut substate_changes = Vec::new();
+        for ((node_id, module_id), node_module_updates) in system_updates {
+            for (substate_key, update) in node_module_updates {
+                let partition_key = D::to_db_partition_key(node_id, *module_id);
+                let sort_key = D::to_db_sort_key(substate_key);
+                let change_action = match update {
+                    DatabaseUpdate::Set(value) => {
+                        match store.get_substate(&partition_key, &sort_key) {
```

Comment:
_(note from author)_

This code is not authored, just moved (out of the `process()` function).
While moving, I noticed that:
- we do this post-tx DB lookups, only because we need to distinguish between an update and create - if the Engine does not distinguish this, maybe we also don't need to?
- we use `system_updates` and have to do `D::to_db_..._key()` - could we use `database_updates` which are already provided by the Engine instead?

Reply:
> we use system_updates and have to do D::to_db_..._key() - could we use database_updates which are already provided by the Engine instead?

Yes - although in a way, I'm not sure why the engine outputs `database_updates` to start with - I feel like we could almost have an `into` from system updates to database updates?

Reply:
> we do this post-tx DB lookups, only because we need to distinguish between an update and create - if the Engine does not distinguish this, maybe we also don't need to?

Ideally we'd like to distinguish them for the Core API, and we'll need to do further work here soon - eg parent map stuff.